# Download and Clean Dataset


In [1]:
import pandas as pd
import numpy as np

## Import data from source and explore it

In [2]:
import ssl

ssl._create_default_https_context = ssl._create_unverified_context

concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So the first concrete sample has 540 cement cubic meter, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa. the first concrete sample is much closer to the second concrete sample

In [3]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [4]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.

### Split data into predictors and target

In [5]:
concrete_data_columns = concrete_data.columns
predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

Check of the predictors and the target dataframes.

In [6]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [7]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

In [8]:
n_cols = predictors.shape[1] # number of predictors
n_cols

8

## Import Keras

In [9]:
import keras

Using TensorFlow backend.


Let's import the rest of the packages from the Keras library that we will need to build our regressoin model.

In [10]:
from keras.models import Sequential
from keras.layers import Dense

Build a Neural Network that has one hidden layer with 10 neurons and a ReLU activation function. It uses the adam optimizer and the mean squared error as the loss function.

In [11]:
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

 Randomly split the data into a training and test sets

In [12]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=42)


# Train and Test the Network

In [13]:
# call the model
model = regression_model()
# fit the model with 50 epochs
epochs = 50
model.fit(X_train, y_train, epochs=epochs, verbose=1)

Epoch 1/50
721/721 [==============================] - 0s 128us/step - loss: 64076.4554
Epoch 2/50
721/721 [==============================] - 0s 26us/step - loss: 12982.3025
Epoch 3/50
721/721 [==============================] - 0s 31us/step - loss: 4773.2188
Epoch 4/50
721/721 [==============================] - 0s 33us/step - loss: 4193.0142
Epoch 5/50
721/721 [==============================] - 0s 26us/step - loss: 3774.5567
Epoch 6/50
721/721 [==============================] - 0s 31us/step - loss: 3385.9545
Epoch 7/50
721/721 [==============================] - 0s 33us/step - loss: 3038.5790
Epoch 8/50
721/721 [==============================] - 0s 38us/step - loss: 2730.2517
Epoch 9/50
721/721 [==============================] - 0s 29us/step - loss: 2427.1434
Epoch 10/50
721/721 [==============================] - 0s 34us/step - loss: 2169.4613
Epoch 11/50
721/721 [==============================] - 0s 33us/step - loss: 1944.3740
Epoch 12/50
721/721 [==============================] - 0s 29

## Evaluating the model on the test data

In [14]:
loss_val = model.evaluate(X_test, y_test)
y_pred = model.predict(X_test)
loss_val

309/309 [==============================] - 0s 51us/step


239.92960444083107

##  Compute the mean squared error between the predicted concrete strength and the actual concrete strength.

In [15]:
from sklearn.metrics import mean_squared_error
mean_square_error = mean_squared_error(y_test, y_pred)
mean = np.mean(mean_square_error)
standard_deviation = np.std(mean_square_error)
print(mean, standard_deviation)

239.92960735764592 0.0


### Create a list of 50 mean squared errors.

In [16]:
total_mean_squared_errors = 50
epochs = 50
mean_squared_errors = []
for i in range(0, total_mean_squared_errors):
    X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mean_square_error = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square_error)

mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("The mean and standard deviation of " +str(total_mean_squared_errors) + " mean squared errors without normalized data. Total number of epochs for each training is: " +str(epochs) + "\n")
print("Mean: "+str(mean))
print("Standard Deviation: "+str(standard_deviation))

MSE 1: 113.55303411884987
MSE 2: 119.12188873784828
MSE 3: 107.58566810089408
MSE 4: 117.42634039326393
MSE 5: 125.89212757098251
MSE 6: 90.82871567315654
MSE 7: 84.50648794822322
MSE 8: 67.10533945768782
MSE 9: 57.334194738888044
MSE 10: 54.79961069347789
MSE 11: 47.180107425331684
MSE 12: 47.67685416138288
MSE 13: 55.248129471220246
MSE 14: 60.90375192882945
MSE 15: 58.722343728766084
MSE 16: 45.45872151272968
MSE 17: 50.83933043248445
MSE 18: 52.64833152795687
MSE 19: 43.804283302578725
MSE 20: 45.57492997499731
MSE 21: 42.27271016438802
MSE 22: 50.278812704734435
MSE 23: 42.32053541905672
MSE 24: 48.279770983847214
MSE 25: 42.735085373171714
MSE 26: 46.02848440620892
MSE 27: 45.36472862515249
MSE 28: 59.62867559044106
MSE 29: 49.05628793216446
MSE 30: 48.293054179466274
MSE 31: 49.560042396718245
MSE 32: 40.108849244596115
MSE 33: 47.110843744864354
MSE 34: 45.14033500893602
MSE 35: 46.331440897821224
MSE 36: 49.99645543638557
MSE 37: 46.976174129251525
MSE 38: 50.03126236998919
MS